_overviews/core/programming-collections-generically.md

```diff
@@ -0,0 +1,320 @@
+---
+layout: singlepage-overview
+title: Programming With Collections Generically (Scala 2.13)
+permalink: /overviews/core/:title.html
+---
+
+**Julien Richard-Foy**
+
+This guide shows how to write operations that can be applied to any collection type and return the same
+collection type, and how to write operations that can be parameterized by the type of collection to build.
+It is recommended to first read the article about the
+[architecture of the collections]({{ site.baseurl }}/overviews/core/architecture-of-scala-213-collections.html).
+
+The following sections present how to **consume**, **produce** and **transform** any collection type.
+
+## Consuming any collection
+
+Several solutions can apply, depending on the desired level of genericity.
+
+### Consuming any *actual* collection
+
+Let’s start with the simplest case: consuming any collection that can be traversed.
+You don’t need to know the precise type of the collection,
+but just that it *is* a collection. This can be achieved by taking an `IterableOnce[A]`
+as parameter, or an `Iterable[A]` if you need more than one traversals. Here is an
+example that shows how to implement a `sumBy` operation that sums the elements of a
+collection after they have been transformed by a function:
+
+~~~ scala
+implicit class SumByOperation[A](coll: IterableOnce[A]) {
+  def sumBy[B](f: A => B)(implicit num: Numeric[B]): B = {
+    val it = coll.iterator
+    var result = f(it.next())
+    while (it.hasNext()) {
+      result = num.plus(result, it.next())
+    }
+    result
+  }
+}
+~~~
+
+We define the `sumBy` operation as an [implicit class](/overviews/core/implicit-classes.html) method so that
+it can be applied to a collection as if it was a method of that collection:
+
+~~~ scala
+case class User(name: String, age: Int)
+
+val users = Seq(User("Alice", 22), User("Bob", 20))
+
+println(users.sumBy(_.age)) // “42”
+~~~
+
+Unfortunately, this extension method does not work with values of type `String` and not
+even with `Array`. This is because these types are not part of the Scala collections
+hierarchy.
+
+### Consuming any type that is *like* a collection
+
+If we want the `sumBy` to work on any type that is *like* a collection, such as `String`
+and `Array`, we have to add another indirection level:
```

Comment:
I find myself wondering here why it isn't sufficient that `Array` and `String` are implicitly convertible to collections. I can figure it out, but it might be better to actually spell it out.